asteroid/masknn/attention.py

```diff
@@ -0,0 +1,229 @@
+import torch.nn.functional as F
+import torch.nn as nn
+from torch.nn.modules.activation import MultiheadAttention
+from asteroid.masknn import activations, norms
+import torch
+from asteroid.utils import has_arg
+
+class ImprovedTransformedLayer(nn.Module):
+    def __init__(self, d_model, n_heads, dim_ff, dropout=0., activation="relu", bidirectional=True, norm="gLN"):
+        super(ImprovedTransformedLayer, self).__init__()
+
+        self.mha = MultiheadAttention(d_model, n_heads, dropout=dropout)
+        self.recurrent = nn.LSTM(d_model, dim_ff, bidirectional=bidirectional)
+        self.dropout = nn.Dropout(dropout)
+        ff_inner_dim = 2*dim_ff if bidirectional else dim_ff
+        self.linear = nn.Linear(ff_inner_dim, d_model)
+        self.activation = activations.get(activation)()
+        self.norm_mha = norms.get(norm)(d_model)
+        self.norm_ff = norms.get(norm)(d_model)
+
+    def forward(self, x):
+        x = x.transpose(1, -1)
+        # x is batch, seq_len, channels
+        # self-attention is applied
+        out = self.mha(x, x, x)[0]
+        x = self.dropout(out) + x
+        x = self.norm_mha(x.transpose(1, -1)).transpose(1, -1)
+
+        # lstm is applied
+        out = self.linear(self.dropout(self.activation(self.recurrent(x)[0])))
+        x = self.dropout(out) + x
+        return self.norm_ff(x.transpose(1, -1))
+
+
+class Oladd(nn.Module):
```

Comment:
This looks amazing ! Just the note about the length caching which is cool but a bit dangerous
.

Reply:
Also this will probably need to go somewhere else, right?

Reply:
probably on dsp no ?

Reply:
Yes, it makes sense